### PR TITLE
Opt: split large packet to n*16KB packet

### LINKED ISSUE #79: Send large package safely
**What would you like to be added**:

To send large package safely, getty should split large package(> 16KiB, for example) into 16KiB slices.

**Why is this needed**:

The large package is likely sent failed whose length is more than 8MiB.

### PR DESCRIPTION
**Pull request by Lvnszn**
**What this PR does**:
split large packet to 16KB packet and send one by one for reduce risk when send large packet.
recv will compact small packet in a large one.
**Which issue(s) this PR fixes**:

Fixes #79

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: session.go [hunk 1]
```diff
@@ -418,11 +419,23 @@ func (s *session) WriteBytes(pkg []byte) (int, error) {
 		return 0, ErrSessionClosed
 	}
 
-	lg, err := s.Connection.send(pkg)
+	totalSize := len(pkg)
```
Comment (AlexStocks) [inline session.go:422]: totalSize := len(pkg)
        pkgArray := make([][]byte, 0, totalSize / maxPacketLen + 1)
	writeSize := 0
	for totalSize >= maxPacketLen {
                 pkgArray = append(pkgArray, pkg[writeSize:(writeSize + maxPacketLen)])
		totalSize -= maxPacketLen
		writeSize += maxPacketLen
	}
        pkgArray = append(pkgArray, pkg[writeSize:])
        lg, err := s.Connection.send(pkgArray)
        if err != nil {
	    return writeSize, perrors.Wrapf(err, "s.Connection.Write(pkg len:%d)", len(pkg))
        }

File: session.go [hunk 2]
```diff
@@ -418,10 +419,20 @@ func (s *session) WriteBytes(pkg []byte) (int, error) {
...
+	pkgs := make([][]byte, 0, totalSize/maxPacketLen+1)
+	writeSize := 0
+	for totalSize >= maxPacketLen {
+		pkgs = append(pkgs, pkg[writeSize:(writeSize+maxPacketLen)])
+		totalSize -= maxPacketLen
+		writeSize += maxPacketLen
+	}
+	pkgs = append(pkgs, pkg[writeSize:])
+	lg, err := s.Connection.send(pkgs)
```
Comment (AlexStocks) [inline session.go:431]: u need not to change the codes so quickly. pls test it firstly.

**Comment by AlexStocks** [comment]
pr 我看了，没多大毛病。关键还是发送大包时，要不要加锁的问题。

File: session.go [hunk 3]
```diff
@@ -129,6 +130,7 @@ type session struct {
 	// goroutines sync
 	grNum uatomic.Int32
 	lock  sync.RWMutex
+	wlock sync.Mutex
```
Comment (AlexStocks) [inline session.go:133]: u should use this wlock before every write action/funcs.

**Review by AlexStocks — APPROVED** [review]

File: session.go [hunk 4]
```diff
@@ -418,11 +422,23 @@ func (s *session) WriteBytes(pkg []byte) (int, error) {
...
+	totalSize, writeSize := len(pkg), 0
+	for totalSize >= maxPacketLen {
+		_, err := s.Connection.send(pkg[writeSize:(writeSize + maxPacketLen)])
+		if err != nil {
+			return writeSize, perrors.Wrapf(err, "s.Connection.Write(pkg len:%d)", len(pkg))
+		}
+		totalSize -= maxPacketLen
+		writeSize += maxPacketLen
+	}
+	lg, err := s.Connection.send(pkg[writeSize:])
```
Comment (wongoo) [inline session.go:436]: does it need to send if the `totalSize = 0`?

**Review by georgehao — APPROVED** [review]

File: session.go (hunk 1 above)
Comment (Lvnszn) [reply session.go:425]: address

File: session.go (hunk 4 above)
Comment (Lvnszn) [reply session.go:440]: IMO, the packet needn't to be send. Addressed.

File: session.go (hunk 3 above)
Comment (Lvnszn) [reply session.go:133]: addressed

File: session.go (hunk 4 above)
Comment (wongoo) [reply session.go:440]: @Lvnszn yes, should return if totalSize already decreased to 0

File: session.go [hunk 5]
```diff
@@ -418,11 +422,27 @@ func (s *session) WriteBytes(pkg []byte) (int, error) {
...
 
-	lg, err := s.Connection.send(pkg)
+	totalSize := len(pkg)
+	if totalSize == 0 {
+		return 0, nil
+	}
+	writeSize := 0
+	s.wlock.Lock()
+	defer s.wlock.Unlock()
+	for totalSize >= maxPacketLen {
```
Comment (wenxuwan) [inline session.go:432]: totalSize change to leftPackageSize or packageSize should be more readable?

File: session.go [hunk 6]
```diff
@@ -418,11 +422,27 @@ func (s *session) WriteBytes(pkg []byte) (int, error) {
...
+		writeSize += maxPacketLen
+	}
+	lg, err := s.Connection.send(pkg[writeSize:])
 	if err != nil {
-		return 0, perrors.Wrapf(err, "s.Connection.Write(pkg len:%d)", len(pkg))
+		return writeSize, perrors.Wrapf(err, "s.Connection.Write(pkg len:%d)", len(pkg))
 	}
-	return lg, nil
+
+	return writeSize + lg, nil
```
Comment (wenxuwan) [inline session.go:445]: here can return totalsize directly ?

File: session.go (hunk 5 above)
Comment (Lvnszn) [reply session.go]: addressed.

File: session.go (hunk 6 above)
Comment (Lvnszn) [reply session.go]: add a var for return, PTAL.

File: session.go (hunk 4 above)
Comment (Lvnszn) [reply session.go]: Addressed.

File: session.go [hunk 7]
```diff
@@ -418,11 +422,28 @@ func (s *session) WriteBytes(pkg []byte) (int, error) {
 		return 0, ErrSessionClosed
 	}
 
-	lg, err := s.Connection.send(pkg)
+	leftPackageSize, totalSize, writeSize := len(pkg), len(pkg), 0
+	s.wlock.Lock()
```
Comment (wenxuwan) [inline session.go:426]: In my opinion, the wlock used for protect big package send, so may be u can change the wlock to rwMutex, and only when package size bigger than 16kb, try to get Lock, other times and other places you can use Rlock.

Comment (AlexStocks) [reply session.go:426]: good idea

Comment (Lvnszn) [reply session.go]: address

File: session.go [hunk 8]
```diff
@@ -418,11 +422,34 @@ func (s *session) WriteBytes(pkg []byte) (int, error) {
 		return 0, ErrSessionClosed
 	}
 
-	lg, err := s.Connection.send(pkg)
+	leftPackageSize, totalSize, writeSize := len(pkg), len(pkg), 0
+	if leftPackageSize >= maxPacketLen {
+		s.packetLock.Lock()
```
Comment (wenxuwan) [inline session.go:427]: only  leftPackageSize > maxPacketLen need lock, equal not needed?

Comment (Lvnszn) [reply session.go:427]: address

**Review by wenxuwan — APPROVED** [review]
LGTM

**Pull request merged**